Add Test::FailWarnings

rafork opened the pull request:
Hi, this adds Test::FailWarnings. It's small (5.3KB) and it's great for automatically treating unexpected warnings as a test failure. Technically, it requires v5.8.1 but your 5.8 image is 5.8.9 so the version requirement isn't documented. It's just included the list of modules that are present in all perl versions. I was hoping that you might consider including it.

rafork commented:
There were CI failures on 5 perl versions, but none of them seem to be related to Test::FailWarnings. It installed fine on all perl versions.

The CI failures on 5.24, 5.22, 5.20 were related to:

    Dist-Zilla-PluginBundle-RJBS-5.024

On 5.10, there were some warnings regarding Test::FailWarnings but it installed:

    Data-Section-0.200007| Warning: prerequisite Test::FailWarnings 0 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Capture::Tiny 0.12 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Test::More 0.96 not found. We have 0.92.
    Test-FailWarnings-0.008| Successfully installed distribution

The CI failure was caused by:

    Data-OptList-0.114

On 5.8, it was similar:

    Data-Section-0.200007| Warning: prerequisite Test::FailWarnings 0 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Capture::Tiny 0.12 not found.
    Test-FailWarnings-0.008| Successfully installed distribution

The CI failure was caused by:

    Data-OptList-0.114

rafork commented:
There were CI failures on 5 perl versions, but none of them seem to be related to Test::FailWarnings. It installed fine on all perl versions.

The CI failures on 5.24, 5.22, 5.20 were related to:

    Dist-Zilla-PluginBundle-RJBS-5.024

On 5.10, there were some warnings regarding Test::FailWarnings:

    Data-Section-0.200007| Warning: prerequisite Test::FailWarnings 0 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Capture::Tiny 0.12 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Test::More 0.96 not found. We have 0.92.
    Test-FailWarnings-0.008| Successfully installed distribution

The CI failure was caused by:

    Data-OptList-0.114

On 5.8, it was similar:

    Data-Section-0.200007| Warning: prerequisite Test::FailWarnings 0 not found.
    Test-FailWarnings-0.008| Warning: prerequisite Capture::Tiny 0.12 not found.
    Test-FailWarnings-0.008| Successfully installed distribution

The CI failure was caused by:

    Data-OptList-0.114

...goes and has a look...

The Data-OptList problem is:

    Data-OptList-0.114| Perl v5.12.0 required--this is only v5.8.9, stopped at Makefile.PL line 5.

The previous versions didn't require v5.12.

And the other problem is:

	Dist-Zilla-PluginBundle-RJBS-5.024| Perl v5.26.0 required--this is only v5.20.3, stopped at Makefile.PL line 5.

The previous versions required v5.20.

So it seems that RJBS has recently changed minimum perl version requirements.

I'm updating the pull request to hopefully adjust to this.

rafork commented:
It's better now, but I'm seeing these warnings about something else:

The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

It's probably coming from perl-actions/install-with-cpm@stable. If so, they'll need to do something about it before the end of this month.

oalders commented on cpanfile:
```diff
@@ -77,7 +82,14 @@ if ( "$]" >= 5.020 ) {
     requires 'Dist::Zilla::PluginBundle::Milla';
-    requires 'Dist::Zilla::PluginBundle::RJBS';
+
+    if ( "$>" < 5.026 ) {
```
Why is this `$>` rather than `$]`?

atoomic replied:
indeed, using effective user id here is a mistake
please adjust https://perldoc.perl.org/perlvar#$%3E

rafork replied:
Changed to $]

oalders commented:
Should we be encouraging the use of `Test::FailWarnings`? It hasn't seen a new release in 10 years. My impression was that https://metacpan.org/pod/Test::Warnings and https://metacpan.org/pod/Test::NoWarnings would be better choices.

atoomic commented on cpanfile:
```diff
@@ -52,6 +53,10 @@ requires 'Test::Synopsis';
 requires 'Test::Version';
 requires 'Test::Warnings';
 
+if ( "$]" >= 5.008001 ) {
```
we are building `5.8` but wonder what version of 5.8 we are providing, I bet this check is useless?

rafork replied:
Deleted when I realised the 5.8 image was 5.8.9 (and there was no 5.6 image).

rafork commented:
Hi, Oops. The `$>` was a typo. It's fixed now.

As for Test::NoWarnings, I prefer Test::FailWarnings, because it's easier to use (because you don't need to increment the number of planned tests), and apparently, its output is nicer (I haven't checked that). Also, Test::Warnings, which is already included, claims to be a replacement for Test::NoWarnings, so there's arguably no need for Test::NoWarnings.

Although, if it's commonly used, that alone is a reason to include it. Any module that is commonly used in testing should probably be included. I'm not sure that it should be a question of encouraging (or discouraging) the use of any particular module. I think it should be a question of speeding up tests for users by including more of the modules they are likely to be using in testing, without making the image too large.

A google search for the three modules shows 1.18B results for Test::Warnings, 440K for Test::NoWarnings, and only 7.2K for Test::FailWarnings, so there's maybe a case for adding Test::NoWarnings, but no case for adding Test::FailWarnings, if those google counts reflect usage in perl modules themselves (probably, but determining that for sure would take more effort than it's worth).

I've removed Test::FailWarnings from the pull request, so all that is left is fixing the cpanfile to adapt to RJBS's recent minimum perl version changes.

I'll just continue explicitly installing Test::FailWarnings for my own tests.

I've squashed the commits and just kept this as the message:

    Adapt to recent RJBS minimum perl version changes

P.S. The perl 5.8 version you are using is 5.8.9. I didn't know when I first did it. I think I removed the check later. And then removed the whole thing. :-)

oalders reviewed (approved):
Thanks, @rafork! 🎉
